Source/Provers/SMTLib/SMTLibBatchTheoremProver.cs

```diff
@@ -0,0 +1,318 @@
+      FlushAxioms();
+
+      Push();
+      SendVCAndOptions(descriptiveName, vcString);
+      SendOptimizationRequests();
+
+      FlushLogFile();
+
+      Process.NewProblem(descriptiveName);
+
+      SendCheckSat();
+      Pop();
+
+      FlushLogFile();
+    }
+
+    public override void Reset(VCExpressionGenerator generator)
+    {
+    }
+
+    public override void FullReset(VCExpressionGenerator generator)
+    {
+      if (options.Solver == SolverKind.Z3 || options.Solver == SolverKind.NoOpWithZ3Options)
+      {
+        this.gen = generator;
+        common.Clear();
+        SetupAxiomBuilder(gen);
+        Axioms.Clear();
+        TypeDecls.Clear();
+        AxiomsAreSetup = false;
+        DeclCollector.Reset();
+        NamedAssumes.Clear();
+        UsedNamedAssumes = null;
+      }
+    }
+
+    // TODO: move to base?
+    [NoDefaultContract]
+    public override async Task<Outcome> CheckOutcome(ErrorHandler handler, int errorLimit, CancellationToken cancellationToken)
+    {
+      var result = await CheckOutcomeCore(handler, cancellationToken, errorLimit);
+
+      FlushLogFile();
+
+      return result;
+    }
+
+    [NoDefaultContract]
+    public override async Task<Outcome> CheckOutcomeCore(ErrorHandler handler, CancellationToken cancellationToken,
+      int errorLimit)
+    {
+      if (Process == null || proverErrors.Count > 0) {
+        return Outcome.Undetermined;
+      }
+
+      try {
+        currentErrorHandler = handler;
+        FlushProverWarnings();
+
+        var result = await GetResponse(cancellationToken);
+
+        if (result == Outcome.Invalid) {
+          var labels = CalculatePath(handler.StartingProcId(), errorModel);
+          if (labels.Length == 0) {
+            // Without a path to an error, we don't know what to report
+            result = Outcome.Undetermined;
+          } else {
+            handler.OnModel(labels, errorModel, result);
+          }
+        }
+
+        // Note: batch mode never tries to discover more errors
+
+        FlushLogFile();
+
+        Close();
+
+        return result;
+      } finally {
+        Close();
+        currentErrorHandler = null;
+      }
+    }
+
+    private async Task<Outcome> GetResponse(CancellationToken cancellationToken)
+    {
+      var outcomeSExp = await Process.GetProverResponse().WaitAsync(cancellationToken);
+      var result = ParseOutcome(outcomeSExp, out var wasUnknown);
+
+      var unknownSExp = await Process.GetProverResponse().WaitAsync(cancellationToken);
+      if (wasUnknown) {
+        result = ParseReasonUnknown(unknownSExp, result);
+      }
+
+      var rlimitSExp = await Process.GetProverResponse().WaitAsync(cancellationToken);
+      resourceCount = ParseRCount(rlimitSExp);
+
+      var modelSExp = await Process.GetProverResponse().WaitAsync(cancellationToken);
+      errorModel = ParseErrorModel(modelSExp);
+
+      return result;
+    }
+
+    private string[] CalculatePath(int controlFlowConstant, Model model)
```

Comment:
Put in a note that we could do `(get-value ControlFlow)` instead of `(get-model)`.